interfaces/apparmor/template_vars.go

```diff
@@ -72,38 +69,60 @@ func legacyVariables(appInfo *snap.AppInfo) string {
-// profile. That same content also decides if the profile is enforcing or
-// advisory (complain). This is used to implement developer mode.
-func (b *Backend) aaHeader(appInfo *snap.AppInfo, developerMode bool) []byte {
-	var text string
-	// Use a different template when requested.
-	if len(b.CustomTemplate) == 0 {
-		text = strings.TrimRight(defaultTemplate, "\n}")
-		// TODO: switch to modernVariables() after the default template is
-		// compatible with them.
-		text = strings.Replace(text, "###VAR###\n", legacyVariables(appInfo), 1)
-	} else {
-		text = strings.TrimRight(b.CustomTemplate, "\n}")
-		text = strings.Replace(text, "###VAR###\n", legacyVariables(appInfo), 1)
-	}
+// The line typically looks like this:
+//
+//   profile "snap.http.GET" (attach_disconnected) {
+func writeProfileAttach(buf *bytes.Buffer, appInfo *snap.AppInfo, developerMode bool) {
+	fmt.Fprintf(buf, "profile \"%s\" (attach_disconnected", interfaces.SecurityTag(appInfo))
 	if developerMode {
 		// XXX: This needs to be verified by security team.
-		text = strings.Replace(text, "(attach_disconnected)", "(attach_disconnected,complain)", 1)
+		fmt.Fprintf(buf, ",complain")
+	}
+	fmt.Fprintf(buf, ") {\n")
+}
+
+const (
+	placeholderVar           = "###VAR###"
+	placeholderProfileAttach = "###PROFILEATTACH###"
+	expectedProfileHeader    = "###PROFILEATTACH### (attach_disconnected) {"
+)
+
+// parseTemplate processes an apparmor template
+func parseTemplate(template []byte) (imports, body []byte, err error) {
+	varStart := bytes.Index(template, []byte(placeholderVar))
+	if varStart == -1 {
+		return nil, nil, fmt.Errorf("cannot find placeholder: %q", placeholderVar)
+	}
+	varEndOffset := bytes.IndexRune(template[varStart:], '\n')
+	if varEndOffset == -1 {
+		return nil, nil, fmt.Errorf("missing newline after placeholder: %q", placeholderVar)
+	}
+	profileAttachStart := bytes.Index(template, []byte(placeholderProfileAttach))
+	if profileAttachStart == -1 {
+		return nil, nil, fmt.Errorf("cannot find placeholder: %q", placeholderProfileAttach)
+	}
+	profileAttachEndOffset := bytes.IndexRune(template[profileAttachStart:], '\n')
+	if profileAttachEndOffset == -1 {
+		return nil, nil, fmt.Errorf("missing newline after placeholder: %q", placeholderProfileAttach)
+	}
+	if string(template[profileAttachStart:profileAttachStart+profileAttachEndOffset]) != expectedProfileHeader {
+		return nil, nil, fmt.Errorf("unsupported profile header (wanted: %s)", expectedProfileHeader)
+	}
+	curlyBraceIndex := bytes.LastIndexByte(template, '}')
+	if curlyBraceIndex == -1 {
+		return nil, nil, fmt.Errorf("missing closing curly brace")
 	}
-	text = strings.Replace(text, "###PROFILEATTACH###",
-		fmt.Sprintf("profile \"%s\"", interfaces.SecurityTag(appInfo)), 1)
-	return []byte(text)
+	imports = template[:varStart]
+	// NOTE: +1 gets rid of the newline so that we get just the body
+	body = template[profileAttachStart+profileAttachEndOffset+1 : curlyBraceIndex]
```

Comment:
I'd prefer to restore the original simpler semantics. Just preserve everything that is not the replaced variables themselves instead of assuming a particular format. These are human-edited, and such variables are implicitly understood to be replacements due to the nature of templates everywhere else. If we break those expectations we'll be setting ourselves up for a serious bug soon.